### ISSUE TITLE
[Feature request] Button Morph

### ISSUE THREAD

**Issue by Sivli-Embir** [PEM--/creativepure#1]
As seen here: http://tympanus.net/Development/ButtonComponentMorph/index.html

Not sure if you take requests. I can see if I could bootstrap a PR at some point.

**Comment by PEM--** [comment]
PR are welcome and greatly appreciated. That would be awesome. You need a hand on this or is the code structure clear enough?

**Comment by Sivli-Embir** [comment]
I am not a designer, business logic is my MO, so at the very least a PR from me will need heavy review. That said I will give a go of it and if I get stuck I will query here. I have been meaning to learn stylus for a while anyway.

**Comment by Sivli-Embir** [comment]
Actually I do have a question. All your logic is pure css as far as I can see (which is really nice). But this case is more along the lines of a modal. Would a template make sense in this case?

``` html
{{#ButtonMorph}}
  <div> modal content </div>
{{/ButtonMorph}}
```

**Comment by Sivli-Embir** [comment]
The alternative will be something like:

``` js

    new UIMorphingButton( document.querySelectorAll( '.morph-button' )[0], {
        closeEl : '.icon-close',
        onBeforeOpen : function() {
        },
        onAfterOpen : function() {
        },
        onBeforeClose : function() {
        },
        onAfterClose : function() {
        }
    } );
```

But this can be very tricky to get right with templates being created/destroyed dynamically. Especially on cleanup. I would rather some kind of clean template that new meteor devs can use without rendering confession.

**Comment by PEM--** [comment]
Well, the philosophy of this package is to provide almost nothing else except Stylus files. This way, it should be usable with Blaze as well as with other templating engines such as React or Angular or Web components (who knows).

Thus, I just provide a basic usage **jade** example, just like this one for buttons:

``` jade
span.input
  input(type='email', id='username', name='username')
  label(for='username'): span Your email
```

If some logic is necessary, I also provide a basic **coffee** example:

``` coffee
Template.myForm.rendered = ->
  Meteor.setTimeout ->
    @$ 'input[type=\'email\']'
    .each (idx, el) =>
      $el = @$ el
      $el.addClass 'filled' unless $el.val().length is 0
  , 300
Template.myForm.events
  'focus input[type=\'email\']': (e, t) ->
    (t.$ e.target).addClass 'filled'
  'blur input[type=\'email\']': (e, t) ->
    (t.$ e.target).removeClass 'filled' if e.target.value.length is 0  
```

For this morphing buttons, I would just add a new Stylus file based on [this file](https://github.com/codrops/ButtonComponentMorph/blob/master/css/component.css) and create the most basic example that uses it in its different types of animations.

**Comment by Sivli-Embir** [comment]
Ah sorry habit, people tend to lose it when you use coffee/jade in issues.

I guess that makes sense. I already converted the that file with http://css2stylus.com/. But their may be ways to clean that up. That said I think some level of js will be need to pull of the modal. But I wonder if we can abstract that into two separate components. I will keep reviewing.

On another note. Do you want to keep using coffee or would you consider switching to ES6 via [Babel](https://atmospherejs.com/grigio/babel)? Given that it will be the standard eventually anyway.

**Comment by PEM--** [comment]
Yep, I prefer using Jade, Coffee and... Stylus, obviously :smile: . It's just an unmatched combo for productivity and semantic.

Nice catch for http://css2stylus.com/. I've bookmarked it. That should ease the integration of other niceties from Codrops. Thanks. I haven't put it to the test but I would say that there are some solid optimization that I've done thanks to [nib](https://github.com/tj/nib), [Autoprefixer](https://github.com/jenius/autoprefixer-stylus), [Axis](http://axis.netlify.com/), [Jeet](http://jeet.gs/) and [Typographic](https://github.com/corysimmons/typographic). All these packages are included with [Maxime's Stylus package](https://github.com/mquandalle/meteor-stylus). I've tried to use them at their maximum so that the compilation phase stay as lightweight as possible and the code is kept DRY.

For ES6, as a strong Coffee user, I'm not too found of it. It's still better than raw vanilla ES5. Its syntactic sugar provides some very good advantages but they are all covered by Coffee since a long time.  I can't see the point in putting accolades and parenthesis in our examples. That would just lengthen the examples without providing a better understanding. But, I'm well aware that there is a serious number of opponents against the Coffee case. It almost equals the number of Coffee addicts that are thrilled to see that a package is done in their favorite language. Semantic over readability and people's habits. A long debate :wink:

**Comment by Sivli-Embir** [comment]
Done for the day but I wanted you to see what I had started, see #2. I tried to remove all the useless styles but I am sure I missed some.

Overall I agree, coffee is my preference. That said I have been talked into rewriting a [number of my packages](https://atmospherejs.com/kestanous) to pure JS time and time again. There is a good argument that says we all know JS and Coffee lovers can just use http://js2.coffee/. Which is why I am happy about ES6. It is a nice middle ground. All that said, coffee for the win and use it as long it makes sense.

**Comment by PEM--** [comment]
@Kestanous, Based on our little chat, I've added a small contributing guide. That could help others contributing in our little package.

**Comment by Sivli-Embir** [comment]
Oh nice, missed that by a few seconds. I will review on the morn.

**Comment by PEM--** [comment]
Super cool :clap: I'm checking your PR and keep you posted.

**Comment by Sivli-Embir** [comment]
Partly implemented, needs more options and ether improve the modal or decouple that logic. It would be neat to use this as a page transition.

**Comment by Sivli-Embir** [comment]
Buggy on iOS safari. It opens off center.  
![img_1919](https://cloud.githubusercontent.com/assets/1283884/7301116/8bbd1b0e-e9ae-11e4-98b0-16b82ce26e80.PNG)

**Comment by PEM--** [comment]
Not usable at all... Dawn, that needs some serious rework.
